### PR TITLE
Add filament lookup plugin

### PR DESCRIPTION
**Pull request by HD650**
##### SUMMARY

Add a filament lookup plugin to show the content of process table.
##### ISSUE TYPE

 - New Module Pull Request

##### COMPONENT NAME

filament_lookup
##### ANSIBLE VERSION

```
ansible 2.7.0.dev0 (filament-lookup 8a59823ae5) last updated 2018/06/05 17:31:41 (GMT -400)
  config file = None
  configured module search path = [u'/home/zhikangzhang/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /home/zhikangzhang/workspace/github/hd650/ansible/lib/ansible
  executable location = /home/zhikangzhang/workspace/github/hd650/ansible/bin/ansible
  python version = 2.7.15 (default, May 16 2018, 17:50:09) [GCC 8.1.1 20180502 (Red Hat 8.1.1-1)]
```

### CODE REVIEW

File: lib/ansible/plugins/lookup/filament_lookup.py [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+
+RETURN = """
+content of process table
+"""
+
+
+from ansible.plugins.lookup import LookupBase
+
+try:
+    import subprocess
```
Comment (webknjaz) [inline lib/ansible/plugins/lookup/filament_lookup.py]: Are you sure you want a fallback display if this import statement fails?

Comment (HD650) [reply lib/ansible/plugins/lookup/filament_lookup.py]: Oh, this will cause fallback display? I just see other plugins do this and I copied. What's the right way to import display?

Comment (webknjaz) [reply lib/ansible/plugins/lookup/filament_lookup.py]: The problem I'm trying to spot is that you have wrapped a number of imports here, so if **any** of them fails:
1) next (import) lines in try block **won't be executed at all**
2) control flow will jump into except block, so that you'll have other import which will be executed and then `display` var initialized

After that, all other code in module will get executed and it seems to rely on the fact that all of other imports succeeded.
In case if they failed you'll get a `NameError`, which will state that it doesn't know about `AnsibleError`, for example, and it doesn't exist.

That workaround is designed to provide a fallback for `display` if importing it from `__main__` module fails.
But by wrapping all the imports you're simply masking lots of different errors, which would have to be processed separately.

So the rule of thumb would be to narrow down try block to some concrete piece of code, which you expect to raise an exception. It is considered harmful to "swallow" arbitrary errors without proper handling of corresponding situations.

File: lib/ansible/plugins/lookup/filament_lookup.py [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+    from __main__ import display
+except ImportError:
+    from ansible.utils.display import Display
+    display = Display()
+
+
+def run_command(terms):
+    command = None
+    # if no argument passed, show whole process table
+    if len(terms) is 0:
```
Comment (webknjaz) [inline lib/ansible/plugins/lookup/filament_lookup.py]: This is wrong. `is` operator is an identity comparison.

Under the hood it's `id(len(terms)) == id(0)`, which effectively compares the addresses of two integers in memory in this case.
It only works here because cpython caches some amount of integers for later reuse.
Proper use of it would be `len(terms) == 0`, which loosely corresponds to `len(terms).__eq__(0)`. Or you could do `if not len(terms)`, which would work because `not 0` is `True` (because `bool(0) is False`).

So the rule of thumb is to always use `is` for comparison with `None`, `True` or `False`. In other cases use numeric comparisons (`==`, `!=`, `>`, `<=` etc.), unless specifically needed.

https://docs.python.org/3/reference/datamodel.html
https://stackoverflow.com/a/133024/595220

File: lib/ansible/plugins/lookup/filament_lookup.py [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+        command = "ps aux"
+    # if get an argument, search it in the process table
+    elif len(terms) is 1:
```
Comment (webknjaz) [inline lib/ansible/plugins/lookup/filament_lookup.py]: same

File: lib/ansible/plugins/lookup/filament_lookup.py [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+        command = "ps aux|grep " + str(terms[0])
```
Comment (webknjaz) [inline lib/ansible/plugins/lookup/filament_lookup.py]: Why do you accept a list and then extract a single arg from there anyway? Maybe you could just concatenate all items there?

File: lib/ansible/plugins/lookup/filament_lookup.py [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+    # other condition, raise exception
+    else:
```
Comment (webknjaz) [inline lib/ansible/plugins/lookup/filament_lookup.py]: You'd probably want to do validation as a first thing in the function and then stop caring about checking things. This seems a logical thing to do.

File: lib/ansible/plugins/lookup/filament_lookup.py (hunk 4 above)
Comment (webknjaz) [reply lib/ansible/plugins/lookup/filament_lookup.py]: Oh, and in Python we tend to avoid string concatenation with "+"

File: lib/ansible/plugins/lookup/filament_lookup.py [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+    from ansible.module_utils._text import to_text
+    import epdb
+    from __main__ import display
+except ImportError:
+    from ansible.utils.display import Display
+    display = Display()
+
+
+def run_command(terms):
+    command = None
```
Comment (webknjaz) [inline lib/ansible/plugins/lookup/filament_lookup.py]: I would probably go for an immediate default:
```python
command = "ps aux"
if terms:
  command = "%s | grep %s" % (command, ' '.join(terms))
```

File: lib/ansible/plugins/lookup/filament_lookup.py [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+    from ansible.errors import AnsibleError
+    from ansible.module_utils._text import to_text
+    import epdb
+    from __main__ import display
+except ImportError:
+    from ansible.utils.display import Display
+    display = Display()
+
+
+def run_command(terms):
```
Comment (webknjaz) [inline lib/ansible/plugins/lookup/filament_lookup.py:41]: You could probably use argument unpacking if you tend to accept multiple args:
```python
def run_command(*terms):
    command = 'ps aux'
    if terms:
        command = '%s | grep %s' % (command, ' '.join(terms))

```

https://docs.python.org/3/tutorial/controlflow.html#more-on-defining-functions

Also common way of having optional function args is to default them to `None`:
```python
def run_command(terms=None):
    if terms is None:
        terms = []  # have some reasonable default if you need
```

File: lib/ansible/plugins/lookup/filament_lookup.py [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+    # if get an argument, search it in the process table
+    elif len(terms) is 1:
+        command = "ps aux|grep " + str(terms[0])
+    # other condition, raise exception
+    else:
+        raise AnsibleError("Argument Fault: 1 string argument expect, {0} got.".format(len(terms)))
+    display.vvv(command)
+
+    p = subprocess.Popen(command, shell=True, stdout=subprocess.PIPE, stderr=subprocess.STDOUT)
+    result = p.stdout.read()
```
Comment (webknjaz) [inline lib/ansible/plugins/lookup/filament_lookup.py]: I bet it's possible that this thing raises `IOError`

File: lib/ansible/plugins/lookup/filament_lookup.py [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+        command = "ps aux"
+    # if get an argument, search it in the process table
+    elif len(terms) is 1:
+        command = "ps aux|grep " + str(terms[0])
+    # other condition, raise exception
+    else:
+        raise AnsibleError("Argument Fault: 1 string argument expect, {0} got.".format(len(terms)))
+    display.vvv(command)
+
+    p = subprocess.Popen(command, shell=True, stdout=subprocess.PIPE, stderr=subprocess.STDOUT)
```
Comment (webknjaz) [inline lib/ansible/plugins/lookup/filament_lookup.py]: I guess this might raise an `OSError`

File: lib/ansible/plugins/lookup/filament_lookup.py [hunk 10]
```diff
@@ -0,0 +1,62 @@
...
+    result = p.stdout.read()
+    return to_text(result)
```
Comment (webknjaz) [inline lib/ansible/plugins/lookup/filament_lookup.py]: So refactored I would have a function like this, in case if you want to have **one optional arg for grep**:
```python
from ansible.module_utils import six
...
...
def run_command(needle=None):
    command = "ps aux"
    if needle is not None:
        command = '%s | grep "%s"' % (command, needle)
    display.vvv(command)

    try:
        proc = (
            subprocess.
            Popen(
                command,
                shell=True,
                stdout=subprocess.PIPE,
                stderr=subprocess.STDOUT,
            )
        )
        return to_text(proc.stdout.read())
    except OSError as os_err:
        six.raise_from(AnsibleError("Failed to fetch processes."), os_err)
    except IOError as io_err:
        six.raise_from(AnsibleError("Failed to read subprocess output."), io_err)
```

File: test/units/plugins/lookup/test_filament_lookup.py [hunk 11]
```diff
@@ -0,0 +1,72 @@
...
+# along with Ansible.  If not, see <http://www.gnu.org/licenses/>.
+# Make coding more python3-ish
+from __future__ import (absolute_import, division, print_function)
+__metaclass__ = type
+
+from ansible.errors import AnsibleError
+from ansible.compat.tests import unittest
+from ansible.compat.tests.mock import MagicMock, patch
+import ansible.plugins.lookup.filament_lookup as filament_lookup
+import subprocess
```
Comment (webknjaz) [inline test/units/plugins/lookup/test_filament_lookup.py]: Please note that it's a good practice to separate stdlib imports from own project imports and from third-party lib imports.

The order should be as follows:
1. stdlib imports
2. empty line
3. third-party
4. empty line
5. own

File: test/units/plugins/lookup/test_filament_lookup.py [hunk 12]
```diff
@@ -0,0 +1,72 @@
...
+#
+# Ansible is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with Ansible.  If not, see <http://www.gnu.org/licenses/>.
+# Make coding more python3-ish
+from __future__ import (absolute_import, division, print_function)
```
Comment (webknjaz) [inline test/units/plugins/lookup/test_filament_lookup.py]: usually, we don't wrap imports with parentheses unless we need multiline

File: test/units/plugins/lookup/test_filament_lookup.py [hunk 13]
```diff
@@ -0,0 +1,72 @@
+# -*- coding: utf-8 -*-
+# (c) 2015, Toshio Kuratomi <tkuratomi@ansible.com>
```
Comment (webknjaz) [inline test/units/plugins/lookup/test_filament_lookup.py]: @HD650 don't copy-paste blindly

File: test/units/plugins/lookup/test_filament_lookup.py (hunk 11 above)
Comment (HD650) [reply test/units/plugins/lookup/test_filament_lookup.py]: Ok, and I just want to reply your review of fallback import and it disappeared. So once somebody changes the code, the corresponding review will disappear?

File: test/units/plugins/lookup/test_filament_lookup.py (hunk 13 above)
Comment (HD650) [reply test/units/plugins/lookup/test_filament_lookup.py]: ...., Yes, next time I will read it first.

File: test/units/plugins/lookup/test_filament_lookup.py (hunk 11 above)
Comment (webknjaz) [reply test/units/plugins/lookup/test_filament_lookup.py]: It collapses, but still accessible from the conversation tab.

File: lib/ansible/plugins/lookup/filament_lookup.py (hunk 4 above)
Comment (HD650) [reply lib/ansible/plugins/lookup/filament_lookup.py]: It's because the lookup plugin has no arguments limitation. So the argument the plugin received is already a list. And I put the argument validation code in run_command because it's easy for unit test. So the right way is to have a function do the validation and a function run the command?

File: test/units/plugins/lookup/test_filament_lookup.py (hunk 11 above)
Comment (webknjaz) [reply test/units/plugins/lookup/test_filament_lookup.py]: https://github.com/ansible/ansible/pull/41161#pullrequestreview-126577732 it's here, just click "Show outdated" next to it.

Comment (webknjaz) [reply test/units/plugins/lookup/test_filament_lookup.py]: This thing, right? https://github.com/ansible/ansible/pull/41161#discussion_r193729928

File: lib/ansible/plugins/lookup/filament_lookup.py (hunk 4 above)
Comment (webknjaz) [reply lib/ansible/plugins/lookup/filament_lookup.py]: You invoke it yourself, so do validation in lookup plugin class

File: test/units/plugins/lookup/test_filament_lookup.py (hunk 11 above)
Comment (HD650) [reply test/units/plugins/lookup/test_filament_lookup.py]: Yes, I saw it.

File: lib/ansible/plugins/lookup/filament_lookup.py [hunk 14]
```diff
@@ -36,13 +36,13 @@
 
 
 def run_command(terms):
-    command = None
+    command = "ps aux"
```
Comment (webknjaz) [inline lib/ansible/plugins/lookup/filament_lookup.py]: Please don't mix unrelated changes in one commit. You may use `git commit --patch` for this.

File: lib/ansible/plugins/lookup/filament_lookup.py [hunk 15]
```diff
@@ -0,0 +1,66 @@
...
+    if terms is not None:
+        command = "%s|grep %s" % (command, str(terms[0]))
+    display.vvv(command)
+
+    try:
+        p = subprocess.Popen(command, shell=True, stdout=subprocess.PIPE, stderr=subprocess.STDOUT)
+        result = p.stdout.read()
+    except OSError as os_err:
+        six.raise_from(AnsibleError("Failed to fetch processes."), os_err)
+    except ISError as io_err:
```
Comment (webknjaz) [inline lib/ansible/plugins/lookup/filament_lookup.py]: `OSError`

Comment (webknjaz) [reply lib/ansible/plugins/lookup/filament_lookup.py]: `IOError`

File: test/integration/targets/filament_lookup/tasks/main.yml [hunk 16]
```diff
@@ -0,0 +1,6 @@
+- name: integration test for one arg lookup
+  assert:
+    that: " one_arg_lookup is defined and 'grep' in one_arg_lookup "
```
Comment (webknjaz) [inline test/integration/targets/filament_lookup/tasks/main.yml:3]:
```yaml
  that: >
    one_arg_lookup is defined and 'grep' in one_arg_lookup
```

Comment (webknjaz) [reply test/integration/targets/filament_lookup/tasks/main.yml:3]: https://learnxinyminutes.com/docs/yaml/

Comment (webknjaz) [reply test/integration/targets/filament_lookup/tasks/main.yml:3]: https://stackoverflow.com/a/21699210/595220

File: test/units/plugins/lookup/test_filament_lookup.py [hunk 17]
```diff
@@ -0,0 +1,72 @@
...
+#
+# You should have received a copy of the GNU General Public License
+# along with Ansible.  If not, see <http://www.gnu.org/licenses/>.
+# Make coding more python3-ish
+from __future__ import (absolute_import, division, print_function)
+__metaclass__ = type
+
+from ansible.errors import AnsibleError
+from ansible.compat.tests import unittest
+from ansible.compat.tests.mock import MagicMock, patch
```
Comment (webknjaz) [inline test/units/plugins/lookup/test_filament_lookup.py]: https://pypi.org/project/pytest-mock/
